File: ios/App/App/NetworkMonitorPlugin.swift
import Foundation
import Capacitor
import Network

@objc(NetworkMonitorPlugin)
public class NetworkMonitorPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "NetworkMonitorPlugin"
    public let jsName = "NetworkMonitor"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise)
    ]

    private var pathMonitor: NWPathMonitor?
    private var monitorQueue: DispatchQueue?
    private var lastStatus: [String: Any]?

    override public func load() {
        setupNetworkMonitoring()
    }

    deinit {
        pathMonitor?.cancel()
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        let status = getCurrentNetworkStatus()
        call.resolve(status)
    }

    private func setupNetworkMonitoring() {
        pathMonitor = NWPathMonitor()
        monitorQueue = DispatchQueue(label: "com.hunteralert.app.networkmonitor")

        pathMonitor?.pathUpdateHandler = { [weak self] path in
            self?.handlePathUpdate(path)
        }

        if let queue = monitorQueue {
            pathMonitor?.start(queue: queue)
        }
    }

    private func handlePathUpdate(_ path: NWPath) {
        let status = getNetworkStatus(from: path)

        // Only notify if status changed
        if let lastStatus = lastStatus,
           NSDictionary(dictionary: lastStatus).isEqual(to: status) {
            return
        }

        lastStatus = status
        notifyListeners("networkStatusChange", data: status)
    }

    private func getCurrentNetworkStatus() -> [String: Any] {
        guard let monitor = pathMonitor,
              let queue = monitorQueue else {
            return getOfflineStatus()
        }

        var currentPath: NWPath?
        queue.sync {
            currentPath = monitor.currentPath
        }

        guard let path = currentPath else {
            return getOfflineStatus()
        }

        return getNetworkStatus(from: path)
    }

    private func getNetworkStatus(from path: NWPath) -> [String: Any] {
        // Check if network is available
        guard path.status == .satisfied else {
            return getOfflineStatus()
        }

        // Determine connectivity type
        var connectivity = "wifi"
        var isSatellite = false

        // Check for satellite (iOS 16.1+)
        if #available(iOS 16.1, *) {
            if path.usesInterfaceType(.other) {
                // In iOS, satellite networks typically appear as "other"
                // This is a heuristic and may need refinement
                connectivity = "satellite"
                isSatellite = true
            }
        }

        if !isSatellite {
            if path.usesInterfaceType(.cellular) {
                connectivity = "cellular"
            } else if path.usesInterfaceType(.wifi) {
                connectivity = "wifi"
            } else if path.usesInterfaceType(.wiredEthernet) {
                connectivity = "wifi" // Treat ethernet as wifi
            }
        }

        // Check if network is constrained
        // isConstrained indicates a low-data mode or carrier-constrained network
        let constrained = path.isConstrained

        // Check if network is expensive (metered/cellular)
        let expensive = path.isExpensive

        // Ultra-constrained is satellite or both constrained and expensive
        let ultraConstrained = isSatellite || (constrained && expensive)

        return [
            "connectivity": connectivity,
            "constrained": constrained,
            "ultraConstrained": ultraConstrained,
            "expensive": expensive
        ]
    }

    private func getOfflineStatus() -> [String: Any] {
        return [
            "connectivity": "offline",
            "constrained": false,
            "ultraConstrained": false,
            "expensive": false
        ]
    }
}

File: android/app/src/main/java/com/hunteralert/app/NetworkMonitorPlugin.java
package com.hunteralert.app;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Build;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "NetworkMonitor")
public class NetworkMonitorPlugin extends Plugin {
    private ConnectivityManager connectivityManager;
    private ConnectivityManager.NetworkCallback networkCallback;
    private JSObject lastStatus;

    @Override
    public void load() {
        connectivityManager = (ConnectivityManager) getContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        registerNetworkCallback();
    }

    @Override
    protected void handleOnDestroy() {
        if (connectivityManager != null && networkCallback != null) {
            connectivityManager.unregisterNetworkCallback(networkCallback);
        }
        super.handleOnDestroy();
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        JSObject status = getCurrentNetworkStatus();
        call.resolve(status);
    }

    private void registerNetworkCallback() {
        // Build a NetworkRequest that accepts constrained networks
        // This is critical for satellite network support
        NetworkRequest.Builder requestBuilder = new NetworkRequest.Builder();

        // Remove the "not constrained" capability to allow satellite and other constrained networks
        // This aligns with Android's satellite network guidelines
        // Requires API 34+ (minSdkVersion = 34)
        requestBuilder.removeCapability(NetworkCapabilities.NET_CAPABILITY_NOT_BANDWIDTH_CONSTRAINED);

        NetworkRequest request = requestBuilder.build();

        networkCallback = new ConnectivityManager.NetworkCallback() {
            @Override
            public void onAvailable(Network network) {
                notifyNetworkChange();
            }

            @Override
            public void onCapabilitiesChanged(Network network, NetworkCapabilities capabilities) {
                notifyNetworkChange();
            }

            @Override
            public void onLost(Network network) {
                notifyNetworkChange();
            }
        };

        connectivityManager.registerNetworkCallback(request, networkCallback);
    }

    private void notifyNetworkChange() {
        JSObject status = getCurrentNetworkStatus();

        // Only notify if status actually changed
        if (lastStatus == null || !status.toString().equals(lastStatus.toString())) {
            lastStatus = status;
            notifyListeners("networkStatusChange", status);
        }
    }

    private JSObject getCurrentNetworkStatus() {
        JSObject status = new JSObject();

        if (connectivityManager == null) {
            status.put("connectivity", "offline");
            status.put("constrained", false);
            status.put("ultraConstrained", false);
            status.put("expensive", false);
            return status;
        }

        Network activeNetwork = connectivityManager.getActiveNetwork();

        if (activeNetwork == null) {
            status.put("connectivity", "offline");
            status.put("constrained", false);
            status.put("ultraConstrained", false);
            status.put("expensive", false);
            return status;
        }

        NetworkCapabilities capabilities = connectivityManager.getNetworkCapabilities(activeNetwork);

        if (capabilities == null) {
            status.put("connectivity", "offline");
            status.put("constrained", false);
            status.put("ultraConstrained", false);
            status.put("expensive", false);
            return status;
        }

        // Determine connectivity type
        String connectivity = "wifi";
        boolean isSatellite = false;

        // Check for satellite transport (API 31+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_SATELLITE)) {
                connectivity = "satellite";
                isSatellite = true;
            }
        }

        if (!isSatellite) {
            if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR)) {
                connectivity = "cellular";
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)) {
                connectivity = "wifi";
            } else if (capabilities.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET)) {
                connectivity = "wifi"; // Treat ethernet as wifi for our purposes
            }
        }

        // Determine if network is constrained
        // A network is constrained if it does NOT have the NOT_BANDWIDTH_CONSTRAINED capability
        // Requires API 34+ (minSdkVersion = 34)
        boolean constrained = !capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_BANDWIDTH_CONSTRAINED);

        // Determine if network is expensive (metered)
        boolean expensive = !capabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED);

        // Ultra-constrained is satellite or both constrained and expensive
        boolean ultraConstrained = isSatellite || (constrained && expensive);

        status.put("connectivity", connectivity);
        status.put("constrained", constrained);
        status.put("ultraConstrained", ultraConstrained);
        status.put("expensive", expensive);

        return status;
    }
}
